### ISSUE TITLE
some unmanaged tools break within interactive rez-env'd shell

### ISSUE THREAD

**Issue by nerdvegas** [nerdvegas/rez#202]
For example, currently xdg-open breaks for me, which causes 'rez-context -g' to fail.

We need a general way to use a tool within a resolved environment if it is available in that env, otherwise fall back to the unresolved env. To do this, interactive rez-env sessions will have to store the pre-resolved env in a file - let's say, 'orig_environ.yaml' alongside the other tempfiles. Then, tools such as xdg-open could be run within this env, if not provided as a package tool in the rez'd env.

**Comment by mstreatfield** [comment]
We've noticed this as well.  In our case meld is a good example.  We only
had a python 2.6 install so it fails in an environment that has resolved to
python 2.7.

Your solution suggests that every package would have to list every tool it
exposes?  Is that right?

**Comment by nerdvegas** [comment]
First, I'm not suggesting that Rez create wrapper scripts on the fly or anything, so that running 'meld' in a rez'd env somehow calls meld in the original (pre-rez) environment. I think that's just going too far - at that point you may as well provide a meld package I think... or, probably better in this case, make 'meld' available in a suite, that way you don't have to include it in the request. I've done this in the past - we have a 'bootstrap' suite which provides a set of tools from outside rez'd environments.

I was considering the cases where Rez itself subproc's programs - such as when 'rez-context -g' is used. In this case the behaviour would be:
1) If the program is listed as a tool by any package in the current context, then run the tool directly within the rez'd env;
2) Otherwise, run the tool in the pre-rez'd env.

Two more things to consider here:

1) It might be useful to provide a new rez tool - 'rez-exec' maybe - that provides the same behaviour from the command line (perhaps with more options). So for eg, you could run "rez-exec meld" within a rez'd env, and it would follow the same rules mentioned above.

2) Another useful feature might be to have a configurable rez setting which, when enabled, causes suites and/or nested rez'd envs to use the pre-rez'd env as the starting point, rather than the current env. In effect this would cause suite tools / further rez'd envs to ignore any changes within the current rez'd env. How could you use this? You might create a 'meld' tool in an empty context within a 'bootstrap' suite that you keep on PATH. If this option were turned on, then you would always have a 'meld' binary available, which would effectively run outside of the current rez env. Of course there's the flipside problem - if you turned on this behaviour globally, there may be cases where you want to use a tool from a suite, but when you run that tool, other changes the user might have made to the current env (unrelated to rez) would disappear. Perhaps the setting would need to be stored on the suite rather than applied globally. A "++use+rez+env" option on the suite tool would be made available also, to override this setting.

**Comment by mstreatfield** [comment]
Ok, that makes more sense!

Yes, `rez exec` sounds useful - although it reminds me of `rez run` which I think got dropped on the switch to rez 2.  `exec` sounds too much like it's going to run in the current environment thought.  `rez revert` (to revert the rez environment) or `rez up` (to go up a sub-shell)?

**Comment by nerdvegas** [comment]
Yeah I'm not fond of 'rez-exec' either. But I don't like 'rez-up' or 'rez-revert' either! Suggestions:
rez-invoke
rez-native-invoke
rez-safe-exec

?
